templates/PlayFab_InstanceApi.cpp.ejs

```diff
@@ -140,39 +158,21 @@ namespace PlayFab
     // Private PlayFabClientInstanceAPI specific
     bool PlayFabClientInstanceAPI::IsClientLoggedIn()
     {
-        return !this->GetOrCreateAuthenticationContext()->clientSessionTicket.empty();
+        return !this->m_context->clientSessionTicket.empty();
     }
 
-    void PlayFabClientInstanceAPI::MultiStepClientLogin(bool needsAttribution)
+    void PlayFabClientInstanceAPI::MultiStepClientLogin(std::shared_ptr<PlayFabAuthenticationContext> context, bool needsAttribution)
```

Comment:
nit: can this shared_ptr be passed by const & instead of copied by value?

Reply:
Let's do it as part of your checkin for this purpose.
I think you're right though, it is a safe candidate here.

Reply:
If you are passing shared_ptr's by const value you should evaluate whether you need to be passing a shared_ptr at all. Does the function being called take shared ownership of the resource? then you should just pass by std::shared_ptr and std::move the parameter whenever ownership needs to be transferred in the function body. If it doesn't need to provide shared_ownership than just pass a raw pointer/reference `Func(const PlayFabAuthenticationContext& context, ...)`

Reply:
Take ownership: No, definitely not. context can be shared by any number of other instanceAPIs.
Because of life-cycle concerns, I'm hesitant to strip the obj out of the shared_ptr. This method could literally be the last ref to the given shared_ptr.
While I know that situation isn't well tested, and I think it'll fail for other reasons, I know for sure it's possible for this here to be the last shared_ptr ref to this obj, and the caller will lose their ref immediately after calling this.
In other words, I think it's worth investigating the case and testing/fixing where this is the last ref (as I'm pretty certain it'll crash), but under that circumstance, I'm pretty certain that this will need to be passed like this, in order to preserve the context obj.

---
In reply to: [343274785](https://github.com/PlayFab/XPlatCoreTemplate/pull/39#discussion_r343274785) [](ancestors = 343274785)

Reply:
Your statement that passing a raw_ptr would drop the reference count is not true (as long as you aren't misusing shared_ptr)

examples:
```
std::shared_ptr<t> sp = ...; // ref count 1
SomeFuncThatTakesSharedPointerByValue(sp); // ref count 2 while in the function, 1 upon returning assuming no one takes additional references inside of the function
```

```
std::shared_ptr<t> sp = ; // ref count 1
SomeFuncThatTakesRawPtr(sp.get()); // ref count stays 1 and the function has no opportunity to accidentally increase the reference count
```

There's a blog by Herb Sutter (one of the c++ big wigs) where he details the when it's appropriate to pass smart pointers around. the TL;DR is unless the callee is manipulating the ownership associated with the resource, pass by raw ptr or ref [https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/](https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/)

---
In reply to: [343327326](https://github.com/PlayFab/XPlatCoreTemplate/pull/39#discussion_r343327326) [](ancestors = 343327326,343274785)

Reply:
We continued this conversation extensively in teams, and Todd & I discussed at length in person.
This is a really hard problem, since we abused shared_pointers to behave pretty close to the C# garbage collector.
It probably shouldn't have been designed this way, but it was, and we copied the C# design directly, rather than having a re-think for the C++ implementation.
This PR doesn't even really change the overall design/usage of shared_ptr regarding contexts & settings, except that it does it "a little more of the same".
So, while I "wontFix" it in this PR, it's not to say it isn't a problem.
At this point a re-think is going to be a customer breaking change, so we should keep that in mind, and schedule it accordingly.
Either way, I'd say it's a different task from this PR, which is just to make the current design work, without data-corruption and leaks.

---
In reply to: [343334818](https://github.com/PlayFab/XPlatCoreTemplate/pull/39#discussion_r343334818) [](ancestors = 343334818,343327326,343274785)